tests/e2e/etcd_release_upgrade_test.go

```diff
@@ -165,3 +170,106 @@ func TestReleaseUpgradeWithRestart(t *testing.T) {
 
 	require.NoError(t, ctlV3Get(cx, []string{kvs[0].key}, []kv{kvs[0]}...))
 }
+
+func TestClusterUpgradeAfterPromotingMembers(t *testing.T) {
+	if !fileutil.Exist(e2e.BinPath.EtcdLastRelease) {
+		t.Skipf("%q does not exist", e2e.BinPath.EtcdLastRelease)
+	}
+
+	e2e.BeforeTest(t)
+
+	currentVersion, err := e2e.GetVersionFromBinary(e2e.BinPath.Etcd)
+	require.NoErrorf(t, err, "failed to get version from binary")
+
+	lastClusterVersion, err := e2e.GetVersionFromBinary(e2e.BinPath.EtcdLastRelease)
+	require.NoErrorf(t, err, "failed to get version from last release binary")
+
+	clusterSize := 3
+
+	epc := createNewClusterByPromotingMembers(t, lastClusterVersion, clusterSize)
+	defer func() {
+		require.NoError(t, epc.Close())
+	}()
+
+	err = e2e.DowngradeUpgradeMembers(t, nil, epc, 3, false, lastClusterVersion, currentVersion)
+	require.NoError(t, err)
+
+	t.Logf("Checking all members' status after upgrading")
+	ensureAllMembersAreVotingMembers(t, epc)
+
+	t.Logf("Checking all members are ready to serve client requests")
+	for i := 0; i < clusterSize; i++ {
+		err = epc.Procs[i].Etcdctl().Put(context.Background(), "foo", "bar", config.PutOptions{})
+		require.NoError(t, err)
+	}
+}
+
+func createNewClusterByPromotingMembers(t *testing.T, clusterVersion *semver.Version, clusterSize int) *e2e.EtcdProcessCluster {
```

Comment:
I think we can minor refactor this function as what I did in https://github.com/etcd-io/etcd/pull/19636.

We can create a separate PR to get it reviewed & merged separately

Reply:
Note: I did some refactoring on this function in https://github.com/etcd-io/etcd/pull/19636, so that it can be reused by my e2e test cases.